Add a create-extension Maven mojo

ppalaga opened the pull request.

ppalaga committed:
Fix #2113 create-extension Maven mojo

ppalaga commented:
cc @emmanuelbernard

ppalaga commented:
Force pushed 1995478 with shorter file paths to make Windows happy.

[This branch](https://github.com/ppalaga/camel-quarkus/commits/190712-create-extension-config) demonstrates how I'd like to use the mojo in camel-quarkus. (A similar setup can be done for the main Quarkus repo too.) 

* The mojo is configured in `extensions/pom.xml`: https://github.com/ppalaga/camel-quarkus/commit/8d8bf565997935dbfdcc50920de3b49b019959a7#diff-93a993e25585d2f32e21d7c8ff6d8242R57
* This configuration contains all settings common to all future extensions in this repo and makes it possible to create new extensions by providing just a single parameter at the command line, e.g.:

```
mvn io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:create-extension -N -Dquarkus.artifactIdBase=log
```

* The result of running the above can be seen in https://github.com/ppalaga/camel-quarkus/commit/33b5992549db40a4af226ead4fb37f89bf05b857

Dufgui commented:
Very interesting feature. Thanks a lot. A really small return (maybe I am wrong):
you code most of the part inside the mojo itself and not put anything in devtools common. But it can be interesting to put in common the template part if someone code the graddle part. And you write directly inside the filesystem. And in this way you cannot write in a zip for exemple. You can reuse/extends ZipProjectWriter and FileProjectWriter no ? What do you think about it ?

ppalaga commented:
> you code most of the part inside the mojo itself and not put anything in devtools common. But it can be interesting to put in common the template part if someone code the graddle part. 

Sorry, I am primarily addressing my own need and this is unfortunately of the scope. Somebody who finds this important is welcome to do that. 

> And you write directly inside the filesystem. And in this way you cannot write in a zip for exemple. You can reuse/extends ZipProjectWriter and FileProjectWriter no ? What do you think about it ?

The current proposal is just solving the task of creating extension modules in existing source trees. How would a zip file be useful in that scenario?

Dufgui commented:
For the put in common ok. It s just a proposal to facility the gradle part. 

I start to work on a web project generator. So I think it’s interesting to do the same on extension. But I am on hold and wait some return to progress again. So it’s up to you for the zip part.

ppalaga commented:
> web project generator. So I think it’s interesting to do the same on extension.

I see, but I guess zipping and being used from an online generator would only make sense for the standalone scenario, that I left unimplemented https://github.com/quarkusio/quarkus/pull/3191/files#diff-e41033b79bb31b53ffe8ec1f25646b7bR476

aloubyansky commented on devtools/maven/pom.xml:
```diff
@@ -102,6 +102,12 @@
+        <dependency>
+            <groupId>org.freemarker</groupId>
+            <artifactId>freemarker</artifactId>
+            <version>2.3.28</version>
```
What do you think about moving the version number to `quarkus-build-parent`'s DM section?

ppalaga replied:
+1

ppalaga replied:
Fixed in 344226776dbd42ba63a4c2163d3ba9955cc232cf

aloubyansky commented on devtools/maven/src/main/java/io/quarkus/maven/CreateExtensionMojo.java:
```diff
@@ -0,0 +1,608 @@
+
+        final Path basePomXml = basedir.resolve("pom.xml");
+        if (Files.exists(basePomXml)) {
+            MavenXpp3Reader reader = new MavenXpp3Reader();
```
It's not a huge deal but for fyi we've been using `io.quarkus.maven.utilities.MojoUtils` to read models and related bits.

ppalaga replied:
There is no `MojoUtils.readPom()` override with an explicit encoding param. Thus -1.

aloubyansky commented on devtools/maven/src/main/resources/create-extension-templates/deployment-pom.xml:
```diff
@@ -0,0 +1,44 @@
+[#if nameBase?? ]    <name>[=namePrefix][=nameBase][=nameSegmentDelimiter]Deployment</name>
+[/#if]
+
+    <dependencies>
```
The deployment artifact should depend on the runtime artifact. This must be documented in the guide.

ppalaga replied:
+1 Actually, I thought, I had it there. Thanks for spotting this!

ppalaga replied:
Fixed in 344226776dbd42ba63a4c2163d3ba9955cc232cf

aloubyansky commented on devtools/maven/src/main/resources/create-extension-templates/deployment-pom.xml:
```diff
@@ -0,0 +1,44 @@
+        <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
```
This assumes the plugin management is inherited from the parent, correct? This either has to be documented or `<version>[=quarkusVersion]</version>` should be added here and in other templates.

ppalaga replied:
> This assumes the plugin management is inherited from the parent, correct? 

Yes, I assume sane projects manage their plugins somewhere up in their module hierarchy.

> This either has to be documented 

Documented where? In the Mojo JavaDoc?

> or `<version>[=quarkusVersion]</version>` should be added here and in other templates.

Well, `<version>[=quarkusVersion]</version>` is not correct for sure. I think if we want to be perfect, we should have a dedicated `compilerPluginVersion` Mojo param. But personally, I'd say this is not worth overengineering at the current stage.

aloubyansky commented:
BTW, it'd be also great to mention this in the guide. And as it's a noteworthy feature, I believe, @gsmet will appreciate a brief summary of it he could include in the release notes.
Thanks @ppalaga!

ppalaga commented:
3442267:

* Rebased on current master
* Addressed some @aloubyansky 's comments

Still open from my PoV:

* How to handle the version of `maven-compiler-plugin`? - If document - where? If version - do we want a separate Mojo param?
* No problem with mentioning the new mojo in the guide, I vote for creating a followup issue and assigning to me.

aloubyansky commented:
@ppalaga there were test failures, could you review those please? Thanks.

ppalaga commented:
0ab6267 should fix the test failures

ppalaga commented:
The CI is passing now

aloubyansky commented on devtools/maven/src/main/resources/create-extension-templates/runtime-pom.xml:
```diff
@@ -0,0 +1,38 @@
+        <plugins>
+            <plugin>
+                <groupId>io.quarkus</groupId>
+                <artifactId>quarkus-bootstrap-maven-plugin</artifactId>
```
Sorry, when I made a comment about the quarkusVersion I meant the version of this plugin, for example, and its configuration. Here it is assumed it is inherited.

ppalaga replied:
Thanks for the explanation, now I see what you mean.

When adding a new extension into an existing tree (the case the current PR implements) I think that assuming the inheritance is the right thing. I guess no sane project wants to repeat the config and version in every of their multiple extension POMs. Documenting it in the mojo JavaDoc is a good idea and I am going to do that.

It is a bit unclear to me how the standalone case (that the current PR is not implementing) should look like. I think regardless whether the standalone scenario will produce a source tree capable of hosting a single or multiple extensions, it will always have some sort of a parent pom where these things can (and should) be set.

Having said that, I believe there is no need to add mojo options for configs/versions of maven plugins. All we need is to document our opinionatedness. WDYT?

aloubyansky replied:
In general, I disagree. If the intention is to provide a tool that will have set the basic project up, meaning to start a new project, then for users who actually start from nothing the current implementation will not go all the way, which surely will be disappointing. I'm not blaming you for not solving all the issues, of course, your contribution is very much appreciated. I am just saying that, imo, it's incomplete.
What it could do in addition, is check whether it is being executed in the context of an existing project and whether the necessary plugin configs can be inherited from the parent POM. If they can't be found then it should add the plugin configs to the POM to be generated.

ppalaga replied:
> What it could do in addition, is check whether it is being executed in the context of an existing project and whether the necessary plugin configs can be inherited from the parent POM. If they can't be found then it should add the plugin configs to the POM to be generated.

I must say the "necessary" is a bit unclear to me. I wonder how the set of "necessary plugin configs" could be defined? Will such a definition stay stable over time? I wonder whether it is possible to implement this in a maintainable way. I am spontaneously skeptic about it but maybe the "necessary configs" are actually easy to define. Could you please give a hint about what is "necessary"?

aloubyansky replied:
Yes, that would actually be the version and config of the quarkus-bootstrap-maven-plugin. As to the stability, the project will generated for a specific version of Quarkus, so it will have to be compatible with that version, of course. It doesn't have to be compatible with older or newer Quarkus versions.

ppalaga replied:
resolved in 52ad079

ppalaga commented:
About  52ad079:

* Added `requiresProject = false` to `@Mojo`
* There is a basic autodetection of `assumeManaged` mojo param that controls whether the versions of deps and config of the bootstrap plugin is added: https://github.com/quarkusio/quarkus/pull/3191/files#diff-e41033b79bb31b53ffe8ec1f25646b7bR373
* Default template for the runtime module changed accordingly https://github.com/quarkusio/quarkus/pull/3191/files#diff-ead5e214bdd42168992f8277004cb379R22
* Added a general purpose utility to transform POM files while keeping the original formatting and comments: https://github.com/quarkusio/quarkus/pull/3191/files#diff-4c8d18606ed61dac84a0bb2870dccd66R36

All issues are solved from my PoV.

ppalaga commented:
f957cc8 should fix the windows build

ppalaga commented:
Is this good to merge now, @aloubyansky ?

aloubyansky commented:
Looks great, thanks a lot @ppalaga !

gsmet commented:
I removed the `noteworthy-feature` label for now because we don't have any documentation about this feature.

We need to have a paragraph in the extension guide explaining how this works.

ppalaga commented:
@gsmet could you please assing https://github.com/quarkusio/quarkus/issues/3386 to me?
